docs/storage/storage.md

```diff
@@ -0,0 +1,186 @@
+through a large tree from the leaves to the root is also bounded. For writes the subtree update
+batches what would be many smaller writes into one larger but manageable one.
+
+We gain space efficiency by not storing intermediate nodes. This is a big saving as it avoids
+storing entire tree levels, which get very large as the tree grows. This is magnified further as we
+store many versions of the tree.
+
+#### Subtree Diagrams
+
+This diagram shows a tree as it might actually be stored by our code using subtrees of depth 8.
+
+Each subtree does not include its "root" node, though this counts as part of the depth. There are
+additional subtrees below and to the right of the child subtree shown, they can't easily be shown
+in the diagram. Obviously, there could be less than 256 "leaf" nodes in the subtrees if they are not
+yet fully populated.
+
+![strata depth 8 tree](StratumDepth8.png, "Stratum Depth 8")
+
+As it's hard to visualize the structure at scale with stratum depth 8 some examples of smaller
+depths might make things clearer, though these are not supported by the code the diagrams are
+much simpler.
+
+This diagram shows a tree with strata depth 2. It is a somewhat special case as all the levels are
+stored. Note that the root node is never stored and is always recalculated.
+
+![strata depth 2 tree diagram](StratumDepth2.png, "Stratum Depth 2")
+
+This diagram shows a tree with strata depth 3. Note that only the bottom level of each subtree is
+stored and how the binary path is used as a subtree prefix to identify subtrees.
+
+![strata depth 3 tree diagram](StratumDepth3.png, "Stratum Depth 3")
+
+### Consistency and Other Requirements
+
+Storage implementations must provide strongly consistent updates to the tree data. Some users may
+see an earlier view than others if updates have not been fully propagated yet but they must not see
+partial updates or inconsistent views.
+
+It is not a requirement that the underlying storage is relational. Our initial implementation uses
+an RDBMS and has this [database schema diagram](database-diagram.pdf, "Trillian DB Schema")
+
+## Map Storage
+
+Need stuff here.
+
+## Log Storage
+
+### The Log Tree
+
+The node tree built for a log is a representation of a Merkle Tree, which starts out empty and grows
+as leaves are added. A Merkle Tree of a specific size is a fixed and well-defined shape.
+                     
+Leaves are never removed and a completely filled in left subtree of the tree
+structure is never further mutated. 
+
+The log stores two hashes per leaf, a raw SHA256 hash of the leaf data used for deduplication and
+the Merkle Leaf Hash of the data, which becomes part of the tree.
+
+### Log NodeIDs / Tree Coordinates
+
+Log nodes are notionally addressed using a three dimensional coordinate tuple (level in tree, index
+in level, revision number).
+
+Level zero is always the leaf level and additional intermediate levels are added above this as the
```

Comment:
There's a danger of confusion having a terminology that includes: 'level' and 'stratum'.  If I was correct above and 'sub-tree' == 'stratum' then clearly stratum != level, but I'm not totally sure.